registry/registry_test.go

```diff
@@ -49,3 +73,563 @@ func TestFindPreviousMajorNotExisting(t *testing.T) {
+	ver.Slug = "app-test"
+	// This version was created yersterday
+	ver.CreatedAt = time.Now().AddDate(0, 0, -1)
+	ver.ID = getVersionID(ver.Slug, ver.Version)
+	err = createVersion(s, db, ver, []*kivik.Attachment{}, app, true)
+	assert.NoError(t, err)
+
+	// Find the versions since last month
+	lastMonth := time.Now().AddDate(0, -1, 0)
+	vers, err := FindLastsVersionsSince(s, "app-test", "stable", lastMonth)
+	assert.NoError(t, err)
+	assert.Equal(t, 1, len(vers))
+	assert.Equal(t, "3.0.0", vers[0].Version)
+}
+
+func TestDeleteVersion(t *testing.T) {
+	s, _ := GetSpace(testSpaceName)
+	// Version 2.0.0 is the only to have an attachment
+	ver, err := findVersion("app-test", "2.0.0", s.VersDB())
+	assert.NoError(t, err)
+	assert.NotNil(t, ver)
+
+	// Check the file is still here
+	conf, err := config.GetConfig()
+	assert.NoError(t, err)
+	sc := conf.SwiftConnection
+
+	var buf = new(bytes.Buffer)
+	prefix := GetPrefixOrDefault(s)
+
+	fp := filepath.Join(ver.Slug, ver.Version, "myfile1")
+	_, err = sc.ObjectGet(prefix, fp, buf, false, nil)
+	assert.NoError(t, err)
+
+	// Delete the version and try to get the (normally) deleted object
+	err = ver.Delete(s)
+	assert.NoError(t, err)
+	_, err = sc.ObjectGet(prefix, fp, buf, false, nil)
+	assert.Equal(t, swift.ObjectNotFound, err)
+
+}
+
+// Download version
+
+func TestDownloadVersioNoManifest(t *testing.T) {
+	missingManifestFile, _ := ioutil.TempFile(os.TempDir(), "cozy-registry-test")
+	tarWriter := tar.NewWriter(missingManifestFile)
+	defer func() {
+		tarWriter.Close()
+		missingManifestFile.Close()
+		os.Remove(missingManifestFile.Name())
+	}()
+	packageContent, err := json.Marshal(defaultPackage())
+	packageHeaders := &tar.Header{
+		Name: "package.json",
+		Size: int64(len(packageContent)),
+		Mode: 777,
+	}
+
+	err = tarWriter.WriteHeader(packageHeaders)
+	assert.NoError(t, err)
+	_, err = tarWriter.Write(packageContent)
+	assert.NoError(t, err)
+	tarWriter.Flush()
+
+	h := sha256.New()
+	fileContent, _ := ioutil.ReadFile(missingManifestFile.Name())
+	_, err = h.Write(fileContent)
+	assert.NoError(t, err)
+
+	// Generating a bad tarball with a missing editor in the manifest
+	opts := &VersionOptions{
+		URL:     "file://" + missingManifestFile.Name(),
+		Sha256:  hex.EncodeToString(h.Sum(nil)),
+		Version: "2.0.0",
+	}
+
+	_, _, err = DownloadVersion(opts)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "does not contain a manifest")
+}
+
+func TestIsValidVersion(t *testing.T) {
+	ver := &VersionOptions{
+		Version: "1.0.0",
+		URL:     "http://foobar.com",
+		Sha256:  "D5AFEAF17396050E17C40E640DBD26DD2B103B5FBC1BB97D3306ED6254322481",
+	}
+	assert.NoError(t, IsValidVersion(ver))
+}
+
+func TestIsValidVersionBadVersion(t *testing.T) {
+	ver := &VersionOptions{
+		Version: "abc",
+		URL:     "",
+		Sha256:  "azerty",
+	}
+	res := IsValidVersion(ver)
+	assert.Error(t, res)
+	assert.Contains(t, res.Error(), "version", "sha256", "url")
+}
+
+func TestMain(m *testing.M) {
```

Comment:
Some CouchDB databases are created for the tests but not deleted at the end. I have some issues to run the test in local because of that, like:

```
--- FAIL: TestCreateApp (0.00s)
    registry_test.go:125: 
        	Error Trace:	registry_test.go:125
        	Error:      	Received unexpected error:
        	            	Application already exists
        	Test:       	TestCreateApp
```